### PR TITLE
Fix  warning: possible null dereference

### LINKED ISSUE #2493: `-fanalyzer` warning: possible null dereference when using TypeAdapters
Branch: `iron`
GCC version: 11.4
OS: Ubuntu 22.04

Steps to reproduce:

build any `rclcpp`-dependent package using TypeAdapters with `-fanalyzer` flag.

See output below:

```
In member function ‘std::unique_ptr<typename rclcpp::TypeAdapter<MessageT>::ros_message_type, typename rclcpp::detail::MessageDeleterHelper<typename rclcpp::TypeAdapter<MessageT>::ros_message_type, AllocatorT>::Deleter> rclcpp::AnySubscriptionCallback<MessageT, AllocatorT>::convert_custom_type_to_ros_message_unique_ptr(const SubscribedType&) [with MessageT = rclcpp::TypeAdapter<StampedOccupancyGrid_CV, nav_msgs::msg::OccupancyGrid_<std::allocator<void> > >; AllocatorT = std::allocator<void>]’:
cc1plus: warning: use of possibly-NULL ‘*this.rclcpp::AnySubscriptionCallback<rclcpp::TypeAdapter<StampedOccupancyGrid_CV, nav_msgs::msg::OccupancyGrid_<std::allocator<void> > >, std::allocator<void> >::ros_message_type_allocator_.__gnu_cxx::new_allocator<nav_msgs::msg::OccupancyGrid_<std::allocator<void> > >::allocate(1, 0)’ where non-null expected [CWE-690] [-Wanalyzer-possible-null-argument]
  ‘std::unique_ptr<typename rclcpp::TypeAdapter<MessageT>::ros_message_type, typename rclcpp::detail::MessageDeleterHelper<typename rclcpp::TypeAdapter<MessageT>::ros_message_type, AllocatorT>::Deleter> rclcpp::AnySubscriptionCallback<MessageT, AllocatorT>::convert_custom_type_to_ros_message_unique_ptr(const SubscribedType&) [with MessageT = rclcpp::TypeAdapter<StampedOccupancyGrid_CV, nav_msgs::msg::OccupancyGrid_<std::allocator<void> > >; AllocatorT = std::allocator<void>]’: event 1
    |
    |/home/user/ros2_iron/install/rclcpp/include/rclcpp/rclcpp/any_subscription_callback.hpp:474:3:
    |  474 |   convert_custom_type_to_ros_message_unique_ptr(const SubscribedType & msg)
    |      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    |      |   |
    |      |   (1) entry to ‘rclcpp::AnySubscriptionCallback<rclcpp::TypeAdapter<StampedOccupancyGrid_CV, nav_msgs::msg::OccupancyGrid_<std::allocator<void> > >, std::allocator<void> >::convert_custom_type_to_ros_message_unique_ptr’
    |
  ‘std::unique_ptr<typename rclcpp::TypeAdapter<MessageT>::ros_message_type, typename rclcpp::detail::MessageDeleterHelper<typename rclcpp::TypeAdapter<MessageT>::ros_message_type, AllocatorT>::Deleter> rclcpp::AnySubscriptionCallback<MessageT, AllocatorT>::convert_custom_type_to_ros_message_unique_ptr(const SubscribedType&) [with MessageT = rclcpp::TypeAdapter<StampedOccupancyGrid_CV, nav_msgs::msg::OccupancyGrid_<std::allocator<void> > >; AllocatorT = std::allocator<void>]’: event 2
    |
    |/usr/include/c++/11/bits/alloc_traits.h:464:28:
    |  464 |       { return __a.allocate(__n); }
    |      |                ~~~~~~~~~~~~^~~~~
    |      |                            |
    |      |                            (2) calling ‘__gnu_cxx::new_allocator<nav_msgs::msg::OccupancyGrid_<std::allocator<void> > >::allocate’ from ‘rclcpp::AnySubscriptionCallback<rclcpp::TypeAdapter<StampedOccupancyGrid_CV, nav_msgs::msg::OccupancyGrid_<std::allocator<void> > >, std::allocator<void> >::convert_custom_type_to_ros_message_unique_ptr’
    |
    +--> ‘_Tp* __gnu_cxx::new_allocator<_Tp>::allocate(__gnu_cxx::new_allocator<_Tp>::size_type, const void*) [with _Tp = nav_msgs::msg::OccupancyGrid_<std::allocator<void> >]’: events 3-6
           |
           |/usr/include/c++/11/ext/new_allocator.h:103:7:
           |  103 |       allocate(size_type __n, const void* = static_cast<const void*>(0))
           |      |       ^~~~~~~~
           |      |       |
           |      |       (3) entry to ‘__gnu_cxx::new_allocator<nav_msgs::msg::OccupancyGrid_<std::allocator<void> > >::allocate’
           |......
           |  111 |         if (__builtin_expect(__n > this->_M_max_size(), false))
           |      |         ~~
           |      |         |
           |      |         (4) following ‘false’ branch...
           |......
           |  127 |         return static_cast<_Tp*>(::operator new(__n * sizeof(_Tp)));
           |      |                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           |      |                                                |
           |      |                                                (5) ...to here
           |      |                                                (6) this call could return NULL
           |
    <------+
    |
  ‘std::unique_ptr<typename rclcpp::TypeAdapter<MessageT>::ros_message_type, typename rclcpp::detail::MessageDeleterHelper<typename rclcpp::TypeAdapter<MessageT>::ros_message_type, AllocatorT>::Deleter> rclcpp::AnySubscriptionCallback<MessageT, AllocatorT>::convert_custom_type_to_ros_message_unique_ptr(const SubscribedType&) [with MessageT = rclcpp::TypeAdapter<StampedOccupancyGrid_CV, nav_msgs::msg::OccupancyGrid_<std::allocator<void> > >; AllocatorT = std::allocator<void>]’: event 7
    |
    |/usr/include/c++/11/bits/alloc_traits.h:464:28:
    |  464 |       { return __a.allocate(__n); }
    |      |                ~~~~~~~~~~~~^~~~~
    |      |                            |
    |      |                            (7) returning to ‘rclcpp::AnySubscriptionCallback<rclcpp::TypeAdapter<StampedOccupancyGrid_CV, nav_msgs::msg::OccupancyGrid_<std::allocator<void> > >, std::allocator<void> >::convert_custom_type_to_ros_message_unique_ptr’ from ‘__gnu_cxx::new_allocator<nav_msgs::msg::OccupancyGrid_<std::allocator<void> > >::allocate’
    |
  ‘std::unique_ptr<typename rclcpp::TypeAdapter<MessageT>::ros_message_type, typename rclcpp::detail::MessageDeleterHelper<typename rclcpp::TypeAdapter<MessageT>::ros_message_type, AllocatorT>::Deleter> rclcpp::AnySubscriptionCallback<MessageT, AllocatorT>::convert_custom_type_to_ros_message_unique_ptr(const SubscribedType&) [with MessageT = rclcpp::TypeAdapter<StampedOccupancyGrid_CV, nav_msgs::msg::OccupancyGrid_<std::allocator<void> > >; AllocatorT = std::allocator<void>]’: event 8
    |
    |cc1plus:
    | (8): argument 'this' (‘*this.rclcpp::AnySubscriptionCallback<rclcpp::TypeAdapter<StampedOccupancyGrid_CV, nav_msgs::msg::OccupancyGrid_<std::allocator<void> > >, std::allocator<void> >::ros_message_type_allocator_.__gnu_cxx::new_allocator<nav_msgs::msg::OccupancyGrid_<std::allocator<void> > >::allocate(1, 0)’) from (6) could be NULL where non-null expected
    |
In file included from /home/user/ros2_iron/install/nav_msgs/include/nav_msgs/nav_msgs/msg/occupancy_grid.hpp:7,
                 from ---omitted---.cpp:15:
/home/user/ros2_iron/install/nav_msgs/include/nav_msgs/nav_msgs/msg/detail/occupancy_grid__struct.hpp:42:12: note: argument 'this' of ‘nav_msgs::msg::OccupancyGrid_<ContainerAllocator>::OccupancyGrid_(rosidl_runtime_cpp::MessageInitialization) [with ContainerAllocator = std::allocator<void>]’ must be non-null
   42 |   explicit OccupancyGrid_(rosidl_runtime_cpp::MessageInitialization _init = rosidl_runtime_cpp::MessageInitialization::ALL)
```

From what I can understand, this is due to:
https://github.com/ros2/rclcpp/blob/04ea0bb00293387791522590b7347a2282cda290/rclcpp/include/rclcpp/any_subscription_callback.hpp#L452
not being checked against `nullptr`.

### PR DESCRIPTION
**Pull request by roncapat**
Closes #2493 
CC @clalancette 

At least for now I drafted a `bad_alloc` exception, but in the same files it is used `std::runtime_error` carrying always an explanatory message. `bad_alloc` by design is not allowed to allocate memory, so it is not designed with error messages in mind.

So I open this as draft so we can devise the best strategy / the one more consistent with `rclcpp` development in general.

### CODE REVIEW

**Review by clalancette — COMMENTED** [review]
I'm OK with this being a `std::bad_alloc`; the truth is that we will almost never run into this situation at runtime, at least on Linux.  @alsora @wjwwood @fujitatomoya Any thoughts here?

@roncapat One thing; you'll have to put a space in between `){` to pass our linters.

File: rclcpp/include/rclcpp/any_subscription_callback.hpp [hunk 1]
```diff
@@ -429,6 +430,9 @@ class AnySubscriptionCallback
     const std::shared_ptr<const ROSMessageType> & message)
   {
     auto ptr = ROSMessageTypeAllocatorTraits::allocate(ros_message_type_allocator_, 1);
+    if (ptr == nullptr){
```
Comment (roncapat) [inline rclcpp/include/rclcpp/any_subscription_callback.hpp:433]:
```suggestion
    if (ptr == nullptr) {
```

File: rclcpp/include/rclcpp/any_subscription_callback.hpp [hunk 2]
```diff
@@ -438,6 +442,9 @@ class AnySubscriptionCallback
     const std::shared_ptr<const rclcpp::SerializedMessage> & serialized_message)
   {
     auto ptr = SerializedMessageAllocatorTraits::allocate(serialized_message_allocator_, 1);
+    if (ptr == nullptr){
```
Comment (roncapat) [inline rclcpp/include/rclcpp/any_subscription_callback.hpp:445]:
```suggestion
    if (ptr == nullptr) {
```

File: rclcpp/include/rclcpp/any_subscription_callback.hpp [hunk 3]
```diff
@@ -451,6 +458,9 @@ class AnySubscriptionCallback
     const std::shared_ptr<const SubscribedType> & message)
   {
     auto ptr = SubscribedTypeAllocatorTraits::allocate(subscribed_type_allocator_, 1);
+    if (ptr == nullptr){
```
Comment (roncapat) [inline rclcpp/include/rclcpp/any_subscription_callback.hpp:461]:
```suggestion
    if (ptr == nullptr) {
```

File: rclcpp/include/rclcpp/any_subscription_callback.hpp [hunk 4]
```diff
@@ -460,6 +470,9 @@ class AnySubscriptionCallback
   {
     if constexpr (rclcpp::TypeAdapter<MessageT>::is_specialized::value) {
       auto ptr = SubscribedTypeAllocatorTraits::allocate(subscribed_type_allocator_, 1);
+      if (ptr == nullptr){
```
Comment (roncapat) [inline rclcpp/include/rclcpp/any_subscription_callback.hpp:473]:
```suggestion
      if (ptr == nullptr) {
```

File: rclcpp/include/rclcpp/any_subscription_callback.hpp [hunk 5]
```diff
@@ -475,6 +488,9 @@ class AnySubscriptionCallback
   {
     if constexpr (rclcpp::TypeAdapter<MessageT>::is_specialized::value) {
       auto ptr = ROSMessageTypeAllocatorTraits::allocate(ros_message_type_allocator_, 1);
+      if (ptr == nullptr){
```
Comment (roncapat) [inline rclcpp/include/rclcpp/any_subscription_callback.hpp:491]:
```suggestion
      if (ptr == nullptr) {
```

File: rclcpp/include/rclcpp/experimental/subscription_intra_process_buffer.hpp [hunk 6]
```diff
@@ -111,6 +111,9 @@ class SubscriptionIntraProcessBuffer : public SubscriptionROSMsgIntraProcessBuff
   {
     if constexpr (!std::is_same<SubscribedType, ROSMessageType>::value) {
       auto ptr = SubscribedTypeAllocatorTraits::allocate(subscribed_type_allocator_, 1);
+      if (ptr == nullptr){
```
Comment (roncapat) [inline rclcpp/include/rclcpp/experimental/subscription_intra_process_buffer.hpp:114]:
```suggestion
      if (ptr == nullptr) {
```

**Comment by jmachowinski** [comment]
Can we move the check into the allocate functions ?

**Comment by alsora** [comment]
> Can we move the check into the allocate functions ?

+1 on @jmachowinski suggestion

I think it's a good change and it could be worth to apply it automatically.

@roncapat note that the Jazzy feature freeze is on Monday.
I think is worth to push this even if it's a API break, and we can document it in the changelog.

**Review by alsora — APPROVED** [review]

**Comment by alsora** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/20844/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/15147/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/21530/)

**Comment by roncapat** [comment]
I don't understand these CI failures. Anyway, sorry for being late on this, I'll try today to do an alternative PR with fixes inside the executor as suggested.

**Comment by roncapat** [comment]
I just gave a look to the code, and frankly speaking I don't understand whether there is a possibility to move the check in the allocator. As the default allocator is `std::allocator<void>`, we can't make it throw inside on ourselves (it's a standard library thing). Aside from all the templating magic there, what seems to happen under the hood is a call to `std::allocator_traits<Alloc>::allocate` passing the allocator, which by default is `std::allocator<void>`.

**Comment by roncapat** [comment]
Please also note that the current patch fixes unhandled `nullptr`s only in some `rclcpp` headers, the ones used by my components per #2493. If there are other `::allocate` calls around `rclcpp` codebase, I believe these would be detected only by a full `-fanalyzer` rebuild of `rclcpp` package. So this fixes `-fanalyzer` builds of end users consuming `rclcpp` API, maybe not `rclcpp` codebase comprehensively.

**Comment by wjwwood** [comment]
`std::allocator<void>::allocate()` already throws `std::bad_alloc`: https://en.cppreference.com/w/cpp/memory/allocator/allocate

As to the CI failures that looks bizarre to me.

**Comment by wjwwood** [comment]
I assume the entry from [asdf](https://gist.githubusercontent.com/alsora/a25f5d4cc933eb86df82f51cc0a9d09b/raw/6579317f4545b42dd007b95e750d1850090af4ec/rclcpp-2496.repos):

```
  ros2/rclcpp:
    type: git
    url: https://github.com/roncapat/rclcpp.git
    version: fix/fanalizer_possible_dereference
```

Points to an out-of-date branch? Maybe it needs to be merged with rolling first? When you provide a custom repos file it doesn't merge (afaik, but I could be wrong about that) with rolling first.

**Comment by roncapat** [comment]
Yes, at least the documentation says so... we may be in front of a GCC bug then? If I'm using standard allocator I shouldn't see this warning popping up.

**Comment by roncapat** [comment]
OK, my fault... [C++ is not officially supported yet](https://gcc.gnu.org/bugzilla/showdependencytree.cgi?id=97110). I leave the bug tracker here because I think it is interesting to have a pointer to current support status and may be very useful in the future.

Closing for now.

**Comment by wjwwood** [comment]
Could be, I'm not familiar with how a function can let the compiler know what it will do. I assume there is an attribute of some sort, maybe that's just missing. I think there's a version of `new` that doesn't throw, and instead returns nullptr, but I think you have to pass `nothrow` to get that. I don't see something equivalent for `allocate()`.

Yeah here: https://en.cppreference.com/w/cpp/memory/new/operator_new#Version_5

**Pull request closed without merging**